File: kluctl/seal/sealer.py
import base64
import hashlib
import logging
import os
import tempfile

from kluctl.utils.kubeseal import kubeseal_fetch_cert, kubeseal_raw
from kluctl.utils.utils import copy_dict, get_tmp_base_dir
from kluctl.utils.yaml_utils import yaml_load_file, yaml_dump, yaml_load

logger = logging.getLogger(__name__)

HASH_ANNOTATION = "kluctl.io/sealedsecret-hashes"

class Sealer:
    def __init__(self, cluster_vars, sealed_secrets_namespace, sealed_secrets_controller_name, force_reseal):
        self.cluster_vars = cluster_vars
        self.force_reseal = force_reseal
        self.cert = kubeseal_fetch_cert(cluster_vars['context'],
                                        controller_ns=sealed_secrets_namespace,
                                        controller_name=sealed_secrets_controller_name)

        # Can't use NamedTemporaryFile here, thanks to windows not allowing to open it in parallel
        self.cert_dir = tempfile.TemporaryDirectory(dir=get_tmp_base_dir())
        self.cert_file = os.path.join(self.cert_dir.name, "cert")
        with open(self.cert_file, "wb") as f:
            f.write(self.cert)

    def do_hash(self, key, secret, secret_name, secret_namespace, scope):
        salt = f"{self.cluster_vars['name']}-{secret_name}-{secret_namespace or '*'}-{key}"
        if scope != "strict":
            salt += f"-{scope}"
        h = hashlib.scrypt(password=secret, salt=salt.encode("utf-8"), n=16384, r=8, p=1)
        return h.hex()

    def encrypt_secret(self, secret, secret_name, secret_namespace, scope):
        return kubeseal_raw(self.cert_file, secret, secret_name, secret_namespace, scope)

    def seal_file(self, path, target_file):
        basename = os.path.basename(target_file)
        os.makedirs(os.path.dirname(target_file), exist_ok=True)

        content = yaml_load_file(path)
        annotations = content.get("metadata", {}).get("annotations", {})
        secret_name = content.get("metadata", {}).get("name")
        secret_namespace = content.get("metadata", {}).get("namespace", "default")
        secret_type = content.get("type", "Opaque")
        scope = "strict"
        if annotations.get("sealedsecrets.bitnami.com/namespace-wide") == "true":
            scope = "namespace-wide"
        elif annotations.get("sealedsecrets.bitnami.com/cluster-wide") == "true":
            scope = "cluster-wide"
        elif "sealedsecrets.bitnami.com/scope" in annotations:
            scope = annotations["sealedsecrets.bitnami.com/scope"]

        existing_content = {}
        existing_hashes = {}
        if os.path.exists(target_file):
            existing_content = yaml_load_file(target_file)
            existing_annotations = existing_content.get("metadata", {}).get("annotations", {})
            if HASH_ANNOTATION in existing_annotations:
                existing_hashes = yaml_load(existing_annotations[HASH_ANNOTATION])

        secrets = {}
        for k, v in content.get("data", {}).items():
            secrets[k] = base64.b64decode(v)
        for k, v in content.get("stringData", {}).items():
            secrets[k] = v.encode("utf-8")

        result_secret_hashes = {}
        result_encrypted_secrets = {}
        result = {
            "apiVersion": "bitnami.com/v1alpha1",
            "kind": "SealedSecret",
            "metadata": {
                "name": secret_name,
                "annotations": {
                    "sealedsecrets.bitnami.com/scope": scope,
                    **({"sealedsecrets.bitnami.com/namespace-wide": "true"} if scope == "namespace-wide" else {}),
                    **({"sealedsecrets.bitnami.com/cluster-wide": "true"} if scope == "cluster-wide" else {}),
                },
            },
            "spec": {
                "template": {
                    "type": secret_type,
                    "metadata": copy_dict(content.get("metadata", {}))
                },
                "encryptedData": result_encrypted_secrets
            }
        }
        if secret_namespace is not None:
            result["metadata"]["namespace"] = secret_namespace
        result["spec"]["template"].pop("data", None)
        result["spec"]["template"].pop("stringData", None)

        changed_keys = []
        for k, v in secrets.items():
            hash = self.do_hash(k, v, secret_name, secret_namespace, scope)
            existing_hash = existing_hashes.get(k, "")
            if hash == existing_hash and not self.force_reseal:
                logger.debug(f"Secret {secret_name} and key {k} is unchanged, skipping encryption")
                result_encrypted_secrets[k] = existing_content.get("spec", {}).get("encryptedData", {}).get(k)
            else:
                logger.debug(f"Secret {secret_name} and key {k} has changed, encrypting it")
                result_encrypted_secrets[k] = self.encrypt_secret(v, secret_name, secret_namespace, scope)
                changed_keys.append(k)
            result_secret_hashes[k] = hash
        for k in existing_hashes.keys():
            if k not in result_encrypted_secrets:
                logger.debug(f"Secret {secret_name} and key {k} has beed deleted")
                changed_keys.append(k)

        result["metadata"]["annotations"][HASH_ANNOTATION] = yaml_dump(result_secret_hashes)

        if result == existing_content:
            logger.info(f"Skipped {basename} as it did not change")
            return

        logger.info(f"Sealed {basename}. New/changed/deleted keys: {', '.join(changed_keys)}")

        with open(target_file, mode='w') as f:
            yaml_dump(result, f)

File: pkg/seal/sealer.go
package seal

import (
	"context"
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"github.com/bitnami-labs/sealed-secrets/pkg/crypto"
	"github.com/kluctl/kluctl/v2/pkg/status"
	"github.com/kluctl/kluctl/v2/pkg/utils"
	"github.com/kluctl/kluctl/v2/pkg/utils/uo"
	"github.com/kluctl/kluctl/v2/pkg/yaml"
	"golang.org/x/crypto/scrypt"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"os"
	"path/filepath"
	"strconv"
)

const hashAnnotation = "kluctl.io/sealedsecret-hashes"
const certHashAnnotation = "kluctl.io/sealedsecret-cert-hash"

type Sealer struct {
	ctx         context.Context
	forceReseal bool
	cert        *x509.Certificate
	pubKey      *rsa.PublicKey
	certHash    string
}

func NewSealer(ctx context.Context, cert *x509.Certificate, forceReseal bool) (*Sealer, error) {
	s := &Sealer{
		ctx:         ctx,
		forceReseal: forceReseal,
		cert:        cert,
	}

	pk, ok := cert.PublicKey.(*rsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("expected RSA public key but found %v", cert.PublicKey)
	}
	s.pubKey = pk

	var err error
	s.certHash, err = HashPublicKey(cert)
	if err != nil {
		return nil, err
	}

	return s, nil
}

func HashSecret(key string, secret []byte, secretName string, secretNamespace string, scope string) string {
	if secretNamespace == "" {
		secretNamespace = "*"
	}
	salt := fmt.Sprintf("%s-%s-%s", secretName, secretNamespace, key)
	if scope != "strict" {
		salt += "-" + scope
	}
	h, err := scrypt.Key(secret, []byte(salt), 1<<14, 8, 1, 64)
	if err != nil {
		panic(err)
	}
	return hex.EncodeToString(h)
}

func encryptionLabel(namespace string, name string, scope string) []byte {
	var l string
	switch scope {
	case "cluster-wide":
		l = ""
	case "namespace-wide":
		l = namespace
	case "strict":
		fallthrough
	default:
		l = fmt.Sprintf("%s/%s", namespace, name)
	}
	return []byte(l)
}

func (s *Sealer) encryptSecret(secret []byte, secretName string, secretNamespace string, scope string) (string, error) {
	// todo
	b, err := crypto.HybridEncrypt(rand.Reader, s.pubKey, secret, encryptionLabel(secretNamespace, secretName, scope))
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(b), nil
}

type sealedSecret struct {
	content  *uo.UnstructuredObject
	hashes   *uo.UnstructuredObject
	certHash string
}

func buildSecretRef(o *uo.UnstructuredObject) string {
	return fmt.Sprintf("%s/%s", o.GetK8sNamespace(), o.GetK8sName())
}

func (s *Sealer) loadExistingSealedSecrets(p string) (map[string]*sealedSecret, error) {
	ret := map[string]*sealedSecret{}

	if !utils.Exists(p) {
		return ret, nil
	}

	list, err := uo.FromFileMulti(p)
	if err != nil {
		return nil, err
	}
	if len(list) != 1 {
		err = nil
	}

	for _, x := range list {
		var ss sealedSecret

		ss.content = x

		a := x.GetK8sAnnotation(hashAnnotation)
		if a != nil {
			ss.hashes, _ = uo.FromString(*a)
		}
		a = x.GetK8sAnnotation(certHashAnnotation)
		if a != nil {
			ss.certHash = *a
		}
		if ss.hashes == nil {
			ss.hashes = uo.New()
		}

		ret[buildSecretRef(x)] = &ss
	}
	return ret, nil
}

func (s *Sealer) SealFile(p string, targetFile string) error {
	err := os.MkdirAll(filepath.Dir(targetFile), 0o700)
	if err != nil {
		return err
	}

	existingSealedSecrets, err := s.loadExistingSealedSecrets(targetFile)
	if err != nil {
		return err
	}

	secrets, err := uo.FromFileMulti(p)
	if err != nil {
		return err
	}

	var result []any

	for _, o := range secrets {
		existing, _ := existingSealedSecrets[buildSecretRef(o)]
		newSealedSecret, err := s.sealSecret(o, existing)
		if err != nil {
			return err
		}
		result = append(result, newSealedSecret.content)
	}

	err = yaml.WriteYamlAllFile(targetFile, result)
	if err != nil {
		return err
	}
	return nil
}

func (s *Sealer) sealSecret(o *uo.UnstructuredObject, existing *sealedSecret) (*sealedSecret, error) {
	secretName := o.ToUnstructured().GetName()
	secretNamespace := o.ToUnstructured().GetNamespace()
	if secretNamespace == "" {
		secretNamespace = "default"
	}
	secretType, ok, err := o.GetNestedString("type")
	if err != nil {
		return nil, err
	}
	if !ok {
		secretType = "Opaque"
	}

	var scope *string
	x, _, _ := o.GetNestedString("metadata", "annotations", "sealedsecrets.bitnami.com/namespace-wide")
	if b, _ := strconv.ParseBool(x); b {
		tmp := "namespace-wide"
		scope = &tmp
	}
	x, _, _ = o.GetNestedString("metadata", "annotations", "sealedsecrets.bitnami.com/cluster-wide")
	if b, _ := strconv.ParseBool(x); b {
		tmp := "cluster-wide"
		scope = &tmp
	}
	if scope == nil {
		x, _, _ = o.GetNestedString("metadata", "annotations", "sealedsecrets.bitnami.com/scope")
		if x == "" {
			x = "strict"
		}
		scope = &x
	}

	secrets := make(map[string][]byte)

	data, ok, err := o.GetNestedObject("data")
	if err != nil {
		return nil, err
	}
	if ok {
		for k, v := range data.Object {
			s, ok := v.(string)
			if !ok {
				return nil, fmt.Errorf("%s is not a string", k)
			}
			secrets[k], err = base64.StdEncoding.DecodeString(s)
			if err != nil {
				return nil, fmt.Errorf("failed to decode base64 string for secret %s and key %s", secretName, k)
			}
		}
	}

	stringData, ok, err := o.GetNestedObject("stringData")
	if err != nil {
		return nil, err
	}
	if ok {
		for k, v := range stringData.Object {
			s, ok := v.(string)
			if !ok {
				return nil, fmt.Errorf("%s is not a string", k)
			}
			secrets[k] = []byte(s)
		}
	}

	resultSecretHashes := make(map[string]string)

	var result sealedSecret
	result.content = uo.New()
	result.content.SetK8sGVK(schema.GroupVersionKind{Group: "bitnami.com", Version: "v1alpha1", Kind: "SealedSecret"})
	result.content.SetK8sName(secretName)
	result.content.SetK8sNamespace(secretNamespace)
	result.content.SetK8sAnnotation("sealedsecrets.bitnami.com/scope", *scope)
	if *scope == "namespace-wide" {
		result.content.SetK8sAnnotation("sealedsecrets.bitnami.com/namespace-wide", "true")
	}
	if *scope == "cluster-wide" {
		result.content.SetK8sAnnotation("sealedsecrets.bitnami.com/cluster-wide", "true")
	}
	_ = result.content.SetNestedField(secretType, "spec", "template", "type")
	metadata, ok, _ := o.GetNestedObject("metadata")
	if ok {
		result.content.SetNestedField(metadata.Object, "spec", "template", "metadata")
	}

	resealAll := false
	if s.forceReseal {
		resealAll = true
		status.Infof(s.ctx, "Forcing reseal of secrets in %s", secretName)
	} else if existing == nil || existing.certHash != s.certHash {
		resealAll = true
		status.Infof(s.ctx, "Cert for secret %s has changed, forcing reseal", secretName)
	}

	for k, v := range secrets {
		hash := HashSecret(k, v, secretName, secretNamespace, *scope)

		var existingHash string
		if existing != nil {
			existingHash, _, _ = existing.hashes.GetNestedString(k)
		}

		doEncrypt := existing == nil || resealAll
		if !doEncrypt && hash != existingHash {
			status.Infof(s.ctx, "Secret %s and key %s has changed, resealing", secretName, k)
			doEncrypt = true
		}

		if !doEncrypt {
			e, ok, _ := existing.content.GetNestedString("spec", "encryptedData", k)
			if ok {
				status.Tracef(s.ctx, "Secret %s and key %s is unchanged", secretName, k)
				result.content.SetNestedField(e, "spec", "encryptedData", k)
				resultSecretHashes[k] = hash
				continue
			} else {
				status.Infof(s.ctx, "Old encrypted secret %s and key %s not found", secretName, k)
				doEncrypt = true
			}
		}

		e, err := s.encryptSecret(v, secretName, secretNamespace, *scope)
		if err != nil {
			return nil, fmt.Errorf("failed to encrypt secret %s with key %s", secretName, k)
		}
		result.content.SetNestedField(e, "spec", "encryptedData", k)
		resultSecretHashes[k] = hash
	}

	resultSecretHashesStr, err := yaml.WriteYamlString(resultSecretHashes)
	if err != nil {
		return nil, err
	}
	result.content.SetK8sAnnotation(hashAnnotation, resultSecretHashesStr)
	result.content.SetK8sAnnotation(certHashAnnotation, s.certHash)

	return &result, nil
}
